### DISCUSSION TITLE
What is `__FnPtr` from `System.Data`?

### DISCUSSION THREAD

**Discussion by sharpninja** [icsharpcode/ILSpy#2433] [General] [answered]
I'm trying to debug System.Data and so I've decompiled it to a project, but I cannot build it because of `__FnPtr` being unknown.  Searching the web hasn't been fruitful.  Where do I find this type defined?

```csharp
      num1 = __calli((__FnPtr<int (IntPtr, _GUID*, void**)>) *(long*) *(long*) pointer2)((void**) iunknownPtr, (_GUID*) ref local1, (IntPtr) ref local2);
```

causes

```
1>    C:\_OneTeam\System.Data\System.Data\NativeOledbWrapper.cs(41,31,41,34): error CS1525: Invalid expression term 'int'
1>    C:\_OneTeam\System.Data\System.Data\NativeOledbWrapper.cs(41,50,41,51): error CS1525: Invalid expression term ','
1>    C:\_OneTeam\System.Data\System.Data\NativeOledbWrapper.cs(41,52,41,56): error CS1525: Invalid expression term 'void'
1>    C:\_OneTeam\System.Data\System.Data\NativeOledbWrapper.cs(41,58,41,59): error CS1525: Invalid expression term ')'
1>    C:\_OneTeam\System.Data\System.Data\NativeOledbWrapper.cs(41,60,41,61): error CS1525: Invalid expression term ')'
1>    C:\_OneTeam\System.Data\System.Data\NativeOledbWrapper.cs(41,121,41,131): error CS1525: Invalid expression term 'ref'
1>    C:\_OneTeam\System.Data\System.Data\NativeOledbWrapper.cs(41,142,41,152): error CS1525: Invalid expression term 'ref'
```

**Comment by dgrunwald** [comment, marked as answer]
Are you sure you were using ILSpy for that decompilation? If so, which version?

`calli` and `FnPtr` refers to [function pointers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers). This has always existed in IL code for Managed C++, but C# only recently (in C# 9) got support for this feature. Current ILSpy versions support the C# 9 feature. Other decompilers that don't support C# 9 might have invented their own non-standard syntax to represent this .NET feature. But your code doesn't look like anything ILSpy would generate, so I suspect you used a different decompiler?

Current ILSpy versions produce the following output for that line in NativeOledbWrapper:
```
num = ((delegate* unmanaged[Stdcall, Stdcall]<IntPtr, _GUID*, void**, int>)(int)(*(uint*)(int)(*(uint*)ptr3)))((nint)ptr3, (_GUID*)System.Runtime.CompilerServices.Unsafe.AsPointer(ref <Module>.IID_IChapteredRowset), (void**)(&ptr2));
```

Note that System.Data seems to have been originally written in C++/CLI, so it won't be possible to perfectly decompile it to C#.
